extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java

```diff
@@ -0,0 +1,782 @@
+              );
+
+              shardIterator = kinesis.getShardIterator(
+                  currRecord.getStream(),
+                  currRecord.getPartitionId(),
+                  ShardIteratorType.AT_SEQUENCE_NUMBER.toString(),
+                  currRecord.getSequenceNumber()
+              ).getShardIterator();
+
+              rescheduleRunnable(recordBufferFullWait);
+              return;
+            }
+          }
+
+          shardIterator = recordsResult.getNextShardIterator(); // will be null if the shard has been closed
+
+          rescheduleRunnable(fetchDelayMillis);
+        }
+        catch (ProvisionedThroughputExceededException e) {
+          log.warn(
+              e,
+              "encounted ProvisionedThroughputExceededException while fetching records, this means "
+              + "that the request rate for the stream is too high, or the requested data is too large for "
+              + "the available throughput. Reduce the frequency or size of your requests."
+          );
+          long retryMs = Math.max(PROVISIONED_THROUGHPUT_EXCEEDED_BACKOFF_MS, fetchDelayMillis);
+          rescheduleRunnable(retryMs);
+        }
+        catch (InterruptedException e) {
+          // may happen if interrupted while BlockingQueue.offer() is waiting
+          log.warn(
+              e,
+              "Interrupted while waiting to add record to buffer, retrying in [%,dms]",
+              EXCEPTION_RETRY_DELAY_MS
+          );
+          if (currRecord != null) {
+            shardIterator = kinesis.getShardIterator(
+                currRecord.getStream(),
+                currRecord.getPartitionId(),
+                ShardIteratorType.AT_SEQUENCE_NUMBER.toString(),
+                currRecord.getSequenceNumber()
+            ).getShardIterator();
+            rescheduleRunnable(EXCEPTION_RETRY_DELAY_MS);
+          } else {
+            throw new ISE("can't reschedule fetch records runnable, current record is null??");
+          }
+        }
+        catch (ExpiredIteratorException e) {
+          log.warn(
+              e,
+              "ShardIterator expired while trying to fetch records, retrying in [%,dms]",
+              fetchDelayMillis
+          );
+          if (recordsResult != null) {
+            shardIterator = recordsResult.getNextShardIterator(); // will be null if the shard has been closed
+            rescheduleRunnable(fetchDelayMillis);
+          } else {
+            throw new ISE("can't reschedule fetch records runnable, recordsResult is null??");
+          }
+        }
+        catch (ResourceNotFoundException | InvalidArgumentException e) {
+          // aws errors
+          log.error(e, "encounted AWS error while attempting to fetch records, will not retry");
+          throw e;
+        }
+        catch (SdkClientException e) {
+          log.warn(e, "encounted unknown AWS exception, retrying in [%,dms]", EXCEPTION_RETRY_DELAY_MS);
+          rescheduleRunnable(EXCEPTION_RETRY_DELAY_MS);
```

Comment:
Probably we can also check this exception is transient or not using `S3Utils.isServiceExceptionRecoverable`.